backend/dal/Services/Admin/Concrete/AgencyService.cs

```diff
@@ -100,35 +100,133 @@ public Agency Get(int id)
             return this.Context.Agencies.Find(id) ?? throw new KeyNotFoundException();
         }
 
+        /// <summary>
+        /// Add a new agency to the datasource.
+        /// </summary>
+        /// <param name="agency"></param>
+        public override void Add(Agency agency)
+        {
+            agency.ThrowIfNull(nameof(agency));
+
+            base.Add(agency);
+            this.Context.Entry(agency).State = EntityState.Detached;
+
+            // If the agency has been added as a sub-agency, then all users who are associated with the parent agency need to be updated.
+            if (agency.ParentId.HasValue)
+            {
+                var users = this.Context.Users.AsNoTracking().Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)).ToArray();
+                users.ForEach(u => agency.Users.Add(new UserAgency(u, agency)));
+            }
+        }
+
         /// <summary>
         /// Updates the specified agency in the datasource.
         /// </summary>
-        /// <param name="entity"></param>
-        /// <exception type="KeyNotFoundException">Entity does not exist in the datasource.</exception>
+        /// <param name="agency"></param>
+        /// <exception type="KeyNotFoundException">agency does not exist in the datasource.</exception>
         /// <returns></returns>
-        public override void Update(Agency entity)
+        public override void Update(Agency agency)
         {
-            entity.ThrowIfNull(nameof(entity));
+            agency.ThrowIfNull(nameof(agency));
+
+            var original = this.Context.Agencies.Find(agency.Id) ?? throw new KeyNotFoundException();
+            var updatedUsers = new List<User>();
+
+            // If the agency has become a sub-agency, or a parent-agency then users will need to be updated.
+            if (original.ParentId.HasValue && !agency.ParentId.HasValue)
+            {
+                // This agency has become a parent agency, all users associated with it will be removed
+                // because users should only be associated with a single parent.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
```

Comment:
I found this comment misleading because updatedUsers are the users that the agency will be added to, but the comment specifically refers to removing associations. I'd prefer it if the comment for a line specifies the functionality of that line. If you put a single line of whitespace between the comment and the new line I'd be fine with this. Or even better, add a comment specific to line 140 and keep the 138-139 comment above that.

Reply:
I just need to provide additional comments for clarification.  The current comment is correct in what needs to be done.  Line 140 is the logic to ensure that the comment is fulfilled.  However the full answer is below.